### DISCUSSION TITLE
Production UPDI programming

### DISCUSSION THREAD

**Discussion by mechatroniks-git** [SpenceKonde/megaTinyCore#476] [Q&A]
I saw this, https://github.com/becky001/pyupdi/blob/master/README.md and wondered if it could be easily adapted for pymcuprog or if it has already been done. I'm not python or r-pi fluent. 

Ideally, looking for a simple bed of nails UPDI programmer with possible board testing functionality that I could provide my board house, one day. 

Or do folks typically pre program the ATtiny chips?

**Comment by SpenceKonde** [comment]
using prog.py from the pymcuprog-derived SerialUPDI would work in a similar manner to pyupdi. Don't even think of using the stock pymcuprog if using serial adapter for UPDI. Actually, it's slow even with their native USB programmers, try programming a 128k file to a Curiosity Nano DA128 vs same file via DxCore's prog.py using FT232RL or CH340G @ 345600 baud. One of those runs at just about the theoretical maximum that the hardware is capable of writing data to the chip at. One of them does not. 

It isn't technically challenging to do UPDI programming with a modern AVR as the programmer (it's even got one-wire serial built in - did you think the UPDI serial interface was anything else than a USART module configured weird by default? 
The challenges are that it's unclear whether there;'s a working SD card library  (I haven't had time to check what, if anything, is standing in the way of this, it seems like we have SPI, so wtf is the hold up).... 

My short-term plan was a keyboard macro to fire off console commands for prog.py....  but I never even got to the that and I bought the hardware last summer. So many exciting things to do so little time,. And then I get stuck wasting that time repeatedly uploading code to batches of dozen+ It would be mighty nice to have some sort of more rigorous solution. 

I do 100% use pogo pins for programming (now that I've finally got the cheap easy to assemble design ready I hope to get them for sale on Tindie soon - delay is time ti takes to write listings and take photos, and then fix the photos so they match the white background on the page and don't look like crap - I am 6 months behind schedule on new products, with mountains of inventory!). programming is fast enough now that I don't waste my time with the clips, I just hold in place. The biggest benefit of a hardware solution would be the ease of integrating a "program button" to kick it off.

**Reply by SpenceKonde** [reply]
Oh, and I don't think I have ever programed one of these prior to soldering it to a PCB.
